armv7a9-zynq7000/timer.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Phoenix-RTOS
+ *
+ * plo - operating system loader
+ *
+ * Triple Timer controler
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+
+#include "../low.h"
+#include "../timer.h"
+
+#include "peripherals.h"
+
+
+/* TODO: this value should be calculated and provide by zynq API */
+#define TTC_SRC_CLK_CPU_1x   111111115 /* Hz */
+#define TTC_DEFAULT_FREQ     1000      /* Hz */
+
+
+enum {
+	clk_ctrl, clk_ctrl2, clk_ctrl3, cnt_ctrl, cnt_ctrl2, cnt_ctrl3, cnt_value, cnt_value2, cnt_value3, interval_val, interval_cnt2, interval_cnt3,
+	match0, match1_cnt2, match1_cnt3, match1, match2_cnt2, match2_cnt3, match2, match3_cnt2, match3_cnt3, isr, irq_reg2, irq_reg3, ier, irq_en2,
+	irq_en3, ev_ctrl_t1, ev_ctrl_t2, ev_ctrl_t3, ev_reg1, ev_reg2, ev_reg3
+};
+
+
+typedef struct {
+	volatile u32 *base;
+	volatile u32 done;
+	u32 ticksPerFreq;
+	u32 leftTime;
```

Comment:
Shouldn't this be volatile?

Reply:
I assumed that only `base` and `done` fields are frequently used so that they can be optimized by the compiler. It is the reason why I added volatile.

I think that it is not necessary to add volatile to the others. What do you think ?

Reply:
in timer.c:128 you poll this value, compiler have full right to optimize this loop to simple infinite loop then `leftTime` is not volatile

Reply:
Done